collections/_article/design-of-the-skeleton-modifier-3d.md

```diff
@@ -0,0 +1,237 @@
+
+Since `AnimationMixer` is applied before the `Skeleton3D` update process, `SkeletonModifier3D` is guaranteed to run after `AnimationMixer`. Also, they do not require `bone_pose_global_override`; This removes any confusion as to whether we should use override or not.
+
+**Here is a SkeletonModifier3D sequence diagram:**
+
+![skeleton modifier process](/storage/blog/design-of-the-skeleton-modifier-3d/skeleton_modifier_process.webp)
+
+Dirty flag resolution may be performed several times per frame, but the update process is a deferred call and is performed only once per frame.
+
+At the beginning of the update process, it stores the pose before the modification process temporarily. When the modification process is complete and applied to the skin, the pose is rolled back to the temporarily stored pose. This performs the role of the past `bone_pose_global_override` which stored the override pose separate from the original pose.
+
+By the way, you may want to get the pose after the modification, or you may wonder why the modifier in the later part cannot enter the original pose when there are multiple modifiers.
+
+We have added some signals for cases where you need a pause at each point in time, so you can use them.
+
+- AnimationMixer: mixer_applied
+	- Notifies when the blending result related have been applied to the target objects
+- SkeletonModifier3D: modification_processed
+	- Notifies when the modification have been finished
+- Skeleton3D: skeleton_updated
+	- Emitted when the final pose has been calculated will be applied to the skin in the update process
+
+Also, note that this process depends on the `Skeleton3D.modifier_callback_mode_process` property.
+
+![modifier callback mode process property](/storage/blog/design-of-the-skeleton-modifier-3d/modifier_callback_mode_process.webp)
+
+For example, in a use case that the node uses the physics process outside of `Skeleton3D` and it affects `SkeletonModifier3D`, the property must be set to `Physics`.
+
+Finally, now we can say that `SkeletonModifier3D` does not make it impossible to do anything that was possible in the past.
+
+# How to make a custom SkeletonModifier3D?
+
+`SkeletonModifier3D` is a virtual class, so you can't add it as stand alone node to a scene.
+
+![add skeleton modifier](/storage/blog/design-of-the-skeleton-modifier-3d/add_skeleton_modifier.webp)
+
+Then, how do we create a custom `SkeletonModifier3D`? Let's try to create a simple custom `SkeletonModifier3D` that points the Y-axis of a bone to a specific coordinate.
+
+## 1. Create a script
+
+Create a blank gdscript file and extends `SkeletonModifier3D`. At this time, register the custom `SkeletonModifier3D` you created with the `class_name` declaration so that it can be added to the scene dock.
```

Comment:
```suggestion
Create a blank gdscript file that extends `SkeletonModifier3D`. At this time, register the custom `SkeletonModifier3D` you created with the `class_name` declaration so that it can be added to the scene dock.
```